lib/proxy.js

```diff
@@ -1,46 +1,7 @@
-/* eslint-disable no-param-reassign */
-
-const urlParse = require('url').parse;
-const ProxyAgent = require('proxy-agent');
-const http = require('http');
-const https = require('https');
-
-const envKeys = Object.keys(process.env).map(key => key.toLocaleUpperCase());
-if (envKeys.includes('HTTP_PROXY') || envKeys.includes('HTTPS_PROXY')) {
-  const proxy = new ProxyAgent();
-  patch(http, 'http', proxy);
-  patch(https, 'https', proxy);
-}
-
-function patch(httpOrHttps, protocol, proxy) {
-  const originalRequest = httpOrHttps.request;
-  httpOrHttps.request = function request(options, callback) {
-    let opts;
-    if (typeof opts === 'string') {
-      opts = urlParse(options);
-    } else {
-      opts = Object.assign({}, options);
-    }
-
-    // Respect the default agent provided by node's lib/https.js
-    if (
-      (opts.agent === null || opts.agent === undefined) &&
-      typeof opts.createConnection !== 'function' &&
-      (opts.host || opts.hostname)
-    ) {
-      opts.agent = proxy;
-    }
-
-    // Set the default port ourselves to prevent Node doing it based on the proxy agent protocol
-    if (
-      opts.protocol === 'https:' ||
-      (!opts.protocol && protocol === 'https')
-    ) {
-      opts.port = opts.port || 443;
-    }
-    if (opts.protocol === 'http:' || (!opts.protocol && protocol === 'http')) {
-      opts.port = opts.port || 80;
-    }
-    return originalRequest.call(httpOrHttps, opts, callback);
-  };
+if (process.env.HTTP_PROXY || process.env.HTTPS_PROXY) {
+  // eslint-disable-next-line global-require
+  require('global-agent/bootstrap');
+  global.GLOBAL_AGENT.HTTP_PROXY = process.env.HTTP_PROXY;
```

Comment:
Can we use `GLOBAL_AGENT_ENVIRONMENT_VARIABLE_NAMESPACE` instead?

From global-agent readme:

![image](https://user-images.githubusercontent.com/6311784/60036938-021bc000-96b9-11e9-9baf-770c1fffa43e.png)

Also I'm wondering if env vars need to be declared before we require, or is after ok?

Reply:
> Can we use GLOBAL_AGENT_ENVIRONMENT_VARIABLE_NAMESPACE instead?

Yes, technically we could but we would then have to fiddle around with the env vars of the executing process. Plus IMO it would be much less obvious what's going on.

> Also I'm wondering if env vars need to be declared before we require, or is after ok?

No, because the current approach just writes to `global.GLOBAL_AGENT` and not to the env vars.